sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/ResourceUtils.java

```diff
@@ -59,54 +59,42 @@ public static String getRole(Protos.Resource resource) {
         return new MesosResource(resource).getRole();
     }
 
-    public static Optional<String> getPrincipal(Protos.Resource resource) {
-        Optional<Protos.Resource.ReservationInfo> reservationInfo = getReservation(resource);
-
-        if (reservationInfo.isPresent()) {
-            return Optional.of(reservationInfo.get().getPrincipal());
-        } else {
-            return Optional.empty();
-        }
-    }
-
     public static Optional<Protos.Resource.ReservationInfo> getReservation(Protos.Resource resource) {
-        if (resource.getReservationsCount() > 0) {
-            return getRefinedReservation(resource);
+        int count = resource.getReservationsCount();
+        if (count > 0) {
+            // This is a refined reservation against a pre-reserved resource. Reservation entries should in this order:
+            // 1. STATIC reservation for the pre-reserved role (e.g. slave_public)
+            // 2. DYNAMIC reservation for our refined role (e.g. slave_public/svc-role)
+            return Optional.of(resource.getReservations(count - 1));
+        } else if (resource.hasReservation()) {
+            // "Classic" reservation against a resource that isn't statically reserved.
+            // This is the common case when reserving resources that aren't pre-reserved.
+            return Optional.of(resource.getReservation());
         } else {
-            return getLegacyReservation(resource);
-        }
-    }
-
-    private static Optional<Protos.Resource.ReservationInfo> getRefinedReservation(Protos.Resource resource) {
-        if (resource.getReservationsCount() == 0) {
+            // No reservations present.
             return Optional.empty();
         }
-
-        return Optional.of(resource.getReservations(resource.getReservationsCount() - 1));
     }
 
-    private static Optional<Protos.Resource.ReservationInfo> getLegacyReservation(Protos.Resource resource) {
-        if (resource.hasReservation()) {
-            return Optional.of(resource.getReservation());
-        } else {
-            return Optional.empty();
-        }
+    public static Optional<String> getPrincipal(Protos.Resource resource) {
```

Comment:
(nit) can simplify this to
```
return getReservation(resource).map(Protos.Resource.ReservationInfo::getPrincipal);
```

Reply:
Aha, I was wondering what the shorthand version was. Fixed here and below. Thanks!